File: 0341 - Flatten Nested List Iterator/solution.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """

class NestedIterator:
    def __init__(self, A: [NestedInteger]):
        self.g = self.gnt(A)
        self.val = None
        self.next()
    
    def next(self) -> int:
        res = self.val
        try:
            self.val = next(self.g)
        except StopIteration:
            self.val = None
        return res
    
    def hasNext(self) -> bool:
        return self.val is not None
        
    def gnt(self, A):
        for i in A:
            if i.isInteger():
                yield i.getInteger()
            else:
                yield from self.gnt(i.getList())
         

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: 0341 - Flatten Nested List Iterator/solution.cpp
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {
 *   public:
 *     // Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     bool isInteger() const;
 *
 *     // Return the single integer that this NestedInteger holds, if it holds a single integer
 *     // The result is undefined if this NestedInteger holds a nested list
 *     int getInteger() const;
 *
 *     // Return the nested list that this NestedInteger holds, if it holds a nested list
 *     // The result is undefined if this NestedInteger holds a single integer
 *     const vector<NestedInteger> &getList() const;
 * };
 */

class NestedIterator {
private:
    stack<NestedInteger> s;
public:
    NestedIterator(vector<NestedInteger> &A) {
        for(auto it = A.rbegin(); it != A.rend(); it++)
            s.push(*it);
    }
    
    int next() {
        int res = s.top().getInteger(); s.pop();
        return res;
    }
    
    bool hasNext() {
        while(!s.empty()){
            if(s.top().isInteger()) return true;
            auto v = s.top().getList(); s.pop();
            for(auto it = v.rbegin(); it != v.rend(); it++)
                s.push(*it);
        }
        return false;
    }
};

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i(nestedList);
 * while (i.hasNext()) cout << i.next();
 */